### PR TITLE
extract title using html tags and metadata

### PR DESCRIPTION
**Pull request by smigad**
the approach is to obtain `<title>` , all `<h1>` tags and `<meta>` with property `og:title`
Priority given to h1 as long as it has at least 75% similarity with either `<title>` or `og:title` if not `og:title` is used if it has at least 75% similarity with `<title>` because respectable sites put proper metadata about the page.
The string similarity metric is Levenshtein distance implemented in util.Str

if everything else above fails and empty title is obtained the final text would be obtained using the previous method.

@akolonin please take a deeper look since I still do not have a good understanding of the entire codebase.

if this is acceptable, #15 will be the next task

### CODE REVIEW

File: src/main/java/net/webstructor/cat/HtmlStripper.java [hunk 1]
```diff
@@ -98,7 +103,7 @@ public static String stripHtmlAnchor(String url){
          */
         public static String attribute(String source, int pos, String name, String baseHref){
 //TODO: lowercase
-        	int till = 1024;
```
Comment (akolonin) [inline src/main/java/net/webstructor/cat/HtmlStripper.java:102]: While fixing formatting is a good thing, too many formatting fixes may get PR more difficult to review...

File: src/main/java/net/webstructor/self/Siter.java [hunk 2]
```diff
@@ -484,17 +486,17 @@ public static int match(Storager storager,Iter iter,ArrayList positions,Thing th
 		//next, if none, create the pattern for the thing name manually
 		if (AL.empty(patterns))
 			//auto-pattern from thing name split apart
-			matches += match(storager,thing.getName(),iter,thing,time,path,positions, thingTexts, thingPaths, imager, linker);
+			matches += match(storager,thing.getName(),iter,thing,time,path,positions, thingTexts, thingPaths, imager, linker, titler);
 		if (!AL.empty(patterns)) {
 			for (Iterator it = patterns.iterator(); it.hasNext();){				
-				matches += match(storager,((Thing)it.next()).getName(),iter,thing,time,path,positions, thingTexts, thingPaths, imager, linker); 
+                matches += match(storager,((Thing)it.next()).getName(),iter,thing,time,path,positions, thingTexts, thingPaths, imager, linker, titler);
```
Comment (akolonin) [inline src/main/java/net/webstructor/self/Siter.java:443]: Formatting glitch?

File: src/main/java/net/webstructor/self/Siter.java [hunk 3]
```diff
@@ -592,8 +594,8 @@ public static boolean readAutoPatterns(Storager storager, Iter iter, Seq patseq,
 	}
 
 	//match one Pattern for one Thing for one Site
-	public static int match(Storager storager,String patstr, Iter iter, Thing thing, Date time, String path, ArrayList positions, MapMap thingTexts, MapMap thingPaths, Imager imager, Imager linker) {
-		Date now = Time.date(time);
```
Comment (akolonin) [inline src/main/java/net/webstructor/self/Siter.java:562]: One more formatting glitch?

File: src/main/java/net/webstructor/self/Siter.java [hunk 4]
```diff
@@ -607,8 +609,10 @@ public static int match(Storager storager,String patstr, Iter iter, Thing thing,
 				break;
 			
 			//plain text before "times" and "is" added
-			String nl_text = summary.toString();
-			
+			String nl_text = titler.getMap(path).get(path).toString();
```
Comment (akolonin) [inline src/main/java/net/webstructor/self/Siter.java]: I don't think that is right. You should add "title" property in addition to "text" property, not replace the "text" with "title", correct?

File: src/main/java/net/webstructor/cat/HtmlStripper.java (hunk 1 above)
Comment (smigad) [reply src/main/java/net/webstructor/cat/HtmlStripper.java:102]: I understand. Sorry about that. Most of them were not intentional. I should've filtered them out before the commit. Will clean up on following commits.

File: src/main/java/net/webstructor/self/Siter.java (hunk 4 above)
Comment (smigad) [reply src/main/java/net/webstructor/self/Siter.java]: :+1: working on it

**Comment by akolonin** [comment]
Well, I think I see the fundamental concern with the suggested "title assignment" approach. Look at the https://www.nytimes.com/ - it has few tens of articles where each article may be source of of a news item. In such case, neither title, nor og:title nor h1 may work, so you need all possible candidates from the html title down to the "text" except. What is suggested in #15 is to use the "spatially closest title candidate preceding the text position", like it is done with images (while the images are looked up for closet before and after). Using Levenshtein difference **may be** (or *may not*) used as an alternative to using spatial difference to select the most similar title candidate preceding the text body by the spacial index instead the most spatially close, but: A) you should measure the different between the "text" and the "title candidate"; B) similarity should it be based on words, instead of letters?

**Pull request merged**